### ISSUE TITLE
Return codes/ Exit statuses -- are they documented ?

### ISSUE THREAD

**Issue by esantoro** [ytdl-org/youtube-dl#882]
Hello there!
I use youtube-dl regularily on my laptop and i've seen it's also used in shk3/edx-downloader (a tool to download videso from edx online university), which I forked tonight and that i started hacking.

Now: one of the bugs i noticed is that it invokes youtube-dl via os.system, but if something bad happens it just lets the download fails and won't re-download the video.

I was then wondering if there are some particular exit codes to check for.
Or can i just assume that anything different from zero is a bad return value ?

Thanks in advance,
Emanuele Santoro

**Comment by jschwalbe** [comment]
+1 (if not already implemented)

**Comment by weedy** [comment]
No they are not. #852
If everything goes well you get 0 most of the time. Everything else is non-zero (1 I think I haven't looked in months).

**Comment by esantoro** [comment]
It might be cool to have some particular return codes at least for some macro-category of errors.

Assuming that `0` is success and everything else is failure, you might want to report when some situations are "recoverable" (the specifically requested format is unavailable? I might want to try and get whatever the best format is).

It's been three years now, and random people still keep coming at this issue. I think it may be worth implementing this feature.

**Comment by yan12125** [comment]
On Linux exit codes can be 0~255. I guess there will eventually be more than 256 cases if we start to define exit codes. Another issue of exit codes is that those numbers are meaningless. I suggest using the Python API instead of relying on process exit codes for automatic jobs. Anyway there should be more exception classes than ```ExtractorError```. A request can be found at #10308.

**Comment by esantoro** [comment]
I said 

> particular return codes at least for some macro-category of errors

@yan12125  said:
> I guess there will eventually be more than 256 cases if we start to define exit codes.

This is why we should have some error codes for macro-categories of errors (network errors, API errors, IO errors etc), and **some** error codes for very common failures (video id unavailable, requested format unavailable, etc).  And a command line flag for detailed failure description.

This way we can have macro- and micro-  error reporting.

**Comment by yan12125** [comment]
Sorry, I didn't read the full thread.

Defining macro categories is a good idea. How about HTTP-like status codes? For example 1~50 indicate network errors, 51~100 indicate I/O errors, and so on (I'm not sure what's "API errors" mean, could you explain more?)

**Comment by esantoro** [comment]
@yan12125  said:
> I'm not sure what's "API errors" mean

API might be interpreted as:

* "you correctly reached the youtube API endpoint but supplied a wrong/insufficient/vague set of arguments" (examples: you asked for a video and supplied a video-id but there is no such video with that id, you asked for 1280x720 mp4 format, but there only is 480x360 mp4 format)

or

* "you called the youtube-dl script with wrong/insufficient command-line options"


The idea here is to let users know **why** youtube-dl failed, so that they can act consequently.

**Comment by kidol** [comment]
I think the long term goal should be that the extractors simply return predefined error codes with optional error messages. These error codes can then be used as exit codes as well.

1-100 = Reserved for network errors, usage errors etc. (everything not extractor related)

101 = Video not found
102 = Video deleted
103 = Country restricted
104 = Copyright claim
105 = Requires login (private video or requires subscription)

**Comment by yan12125** [comment]
101 is already implemented for "--max-downloads exceeded". Sounds like it's some "API error" proposed by @esantoro. It shouldn't be changed, so 10x can be reserved for API errors.

**Comment by esantoro** [comment]
@yan12125 , @kidol we are definitely movin in the right direction here!

**Comment by hz** [comment]
It's been sixt years now, and random people still keep coming at this issue. I think it may be worth implementing this feature.

**Comment by hz** [comment]
It's been six years now, and random people still keep coming at this issue. I think it may be worth implementing this feature.

**Comment by esantoro** [comment]
Yeah I'm honestly sorry I've been bothering the developers for the last six years with this, I really didn't mean to.
I still think it would be worth implementing, though.

**Comment by rag-hav** [comment]
+1 from me too.

**Comment by toppits** [comment]
would like to add my vote as well.

**Comment by esantoro** [comment]
Lol almost seven years and and people are still asking for this.

**Comment by Jokertion** [comment]
+1 That's useful for user.

**Comment by esantoro** [comment]
This issue is officially seven years old! :tada:

**Comment by Helspank** [comment]
+1 documented Exit Codes would be very nice.

**Comment by hejops** [comment]
Eight years now...

**Comment by FelixKainz** [comment]
I am surprised this is not implemented yet. Much needed IMO.

**Comment by maetthew** [comment]
+1 Exit codes would be very helpful

**Comment by richiedevs** [comment]
+1 from me aswell

**Comment by akmalharith** [comment]
I understand there's no proper specific return codes for errors. 

But for now you can try to parse the shell exit code instead. 
``` bash
#!/bin/bash 
download_url=$(youtube-dl -g https://youtube.com/channel/channelwitherror/live)

if [[ $? != 0 ]];
then
    echo "Error retrieving channel live streams."
    echo "Not downloading."
else
    echo "Actually download the streams here."
    curl $download_url > live_streams.m3u8
fi
```

**Comment by vitalash1** [comment]
This remains a major oversight.

**Comment by dirkf** [comment]
https://github.com/ytdl-org/youtube-dl/issues/31167#issuecomment-1214421230

Also, yt-dlp/yt-dlp#457
